compiler/src/scala/quoted/runtime/impl/QuoteMatcher.scala

```diff
@@ -452,20 +433,52 @@ object QuoteMatcher {
       accumulator.apply(Set.empty, term)
   }
 
+  enum MatchResult:
+    case ClosedTree(tree: Tree)
+    case OpenTree(tree: Tree, patternTpe: Type, args: List[Tree], env: Env)
+
+    def toExpr(mapTypeHoles: TypeMap)(using Context): Expr[Any] = this match
+      case MatchResult.ClosedTree(tree) =>
+        new ExprImpl(tree, SpliceScope.getCurrent)
+      case MatchResult.OpenTree(tree, patternTpe, args, env) =>
+        def hoasClosure = {
+          val names: List[TermName] = args.map {
+            case Block(List(DefDef(nme.ANON_FUN, _, _, Apply(Ident(name), _))), _) => name.asTermName
+            case arg => arg.symbol.name.asTermName
+          }
+          val paramTypes = args.map(x => mapTypeHoles(x.tpe.widenTermRefExpr))
+          val methTpe = MethodType(names)(_ => paramTypes, _ => mapTypeHoles(patternTpe))
+          val meth = newAnonFun(ctx.owner, methTpe)
+          def bodyFn(lambdaArgss: List[List[Tree]]): Tree = {
+            val argsMap = args.map(_.symbol).zip(lambdaArgss.head).toMap
```

Comment:
Could use a view to avoid intermediate allocations:
```suggestion
            val argsMap = args.view.map(_.symbol).zip(lambdaArgss.head).toMap
```